### PR TITLE
Make sure Tomcat thread plugin working in Tomcat 9

### PR DESCRIPTION
**Pull request by wu-sheng**
- [ ] If this pull request closes/resolves/fixes an existing issue, replace the issue number. Closes #<issue number>.
- [ ] Update the [`CHANGES` log](https://github.com/apache/skywalking-java/blob/main/CHANGES.md).

@xu1009 We found a cast exception in Tomcat 9. We supported Tomcat 9 as plugin test container, could you add another test?

Link to https://github.com/apache/skywalking-java/pull/105

### CODE REVIEW

**Comment by wu-sheng** [comment]
@xu1009 And I am not sure whether this fix works, `org.apache.tomcat.util.threads.ThreadPoolExecutor`'s hierarchy tree is changed from 8 to 9.

**Comment by wu-sheng** [comment]
This doesn't work.

```
  java.lang.LinkageError: loader (instance of  org/springframework/boot/loader/LaunchedURLClassLoader): attempted  duplicate class definition for name: "org/apache/tomcat/util/threads/ThreadPoolExecutor"
  	at java.lang.ClassLoader.defineClass1(Native Method) ~[?:1.8.0_322]
  	at java.lang.ClassLoader.defineClass(ClassLoader.java:756) ~[?:1.8.0_322]
  	at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142) ~[?:1.8.0_322]
```

**Review by kezhenxu94 — APPROVED** [review]

File: apm-sniffer/apm-sdk-plugin/tomcat-thread-pool-plugin/src/main/java/org/apache/skywalking/apm/plugin/tomcat/thread/pool/TomcatThreadExecutorInterceptor.java [hunk 1]
```diff
@@ -18,37 +18,40 @@
...
 import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.EnhancedInstance;
 import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.InstanceConstructorInterceptor;
+import org.apache.tomcat.util.threads.ThreadPoolExecutor;
 
 public class TomcatThreadExecutorInterceptor implements InstanceConstructorInterceptor {
 
+    private static final String METER_NAME = "thread_pool";
+    private static final String METRIC_POOL_NAME_TAG_NAME = "pool_name";
     private static final String THREAD_POOL_NAME = "tomcat_execute_pool";
-    private static final Map<String, Function<ThreadPoolExecutor, Supplier<Double>>> METRIC_MAP = new HashMap<String, Function<ThreadPoolExecutor, Supplier<Double>>>() {{
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/tomcat-thread-pool-plugin/src/main/java/org/apache/skywalking/apm/plugin/tomcat/thread/pool/TomcatThreadExecutorInterceptor.java:33]: @xu1009 I removed this, but hard code them in the method.(Verified by H2 case adjusted by @kezhenxu94 )

The reason is, `org.apache.tomcat.util.threads.ThreadPoolExecutor` is rewrited between `9.0.21` and `9.0.54`(don't check for more details). It doesn't **extends java.util.concurrent.ThreadPoolExecutor** anymore. Then we faced type casting exception.

image

Ref to 9.0.52, https://github.com/apache/tomcat/blob/9.0.52/java/org/apache/tomcat/util/threads/ThreadPoolExecutor.java

Comment (wu-sheng) [reply apm-sniffer/apm-sdk-plugin/tomcat-thread-pool-plugin/src/main/java/org/apache/skywalking/apm/plugin/tomcat/thread/pool/TomcatThreadExecutorInterceptor.java:33]: If I use  `org.apache.tomcat.util.threads.ThreadPoolExecutor` in this static block, the JVM class loader will fail when loading this interceptor, which leaves me no choice but to move all logic inside the `onConstruct` method.

**Review by kezhenxu94 — APPROVED** [review]

**Comment by wu-sheng** [comment]
> Test / h2-scenario (pull_request) Successful in 1m

This is passed. @kezhenxu94 I think after merge, another showcase update should be able to fix this.

**Pull request merged**